File: typescript/x402.ts
export const BASE_USDC_CONTRACT = '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913';
export const BASE_CHAIN_ID = 8453;
export const EXPECTED_PAYOUT_RECIPIENT = '0x6d6c398390cfb88f1cd42715b84906a0bd6652aa';
export const DEFAULT_MAX_PRICE_USD = 0.05;

export interface X402SignerClientOptions {
  wallet?: any;
  walletSigner?: any;
  baseUrl?: string;
  timeoutMs?: number;
  expectedRecipient?: string;
  maxPriceUsd?: number;
}

export interface X402PaymentChallenge {
  x402Version?: number;
  scheme?: string;
  tokenName?: string;
  tokenVersion?: string;
  assetContract?: string;
  chainId?: number | string;
  payTo?: string;
  recipient?: string;
  amountUnits?: string;
  maxAmountRequired?: string;
  /** x402 v2 nests the offer here. Its presence selects the v2 envelope. */
  accepts?: any[];
  resource?: any;
  asset?: string;
  network?: string;
  extra?: { name?: string; version?: string };
  maxTimeoutSeconds?: number;
  amount?: string;
  price?: string;
  [key: string]: any;
}

/** The canonical x402 v2 payment payload. */
export interface X402SignedAuthorizationV2 {
  x402Version: 2;
  resource?: any;
  /** The server's advertised extensions, echoed back. The CDP Bazaar indexes a
   *  resource from the bazaar extension carried by the settlement. */
  extensions?: any;
  /** The chosen offer echoed back verbatim. The server deep-equality matches
   *  this against its own requirements to learn which offer is being paid. */
  accepted: any;
  payload: {
    authorization: {
      from: string;
      to: string;
      value: string;
      validAfter: string;
      validBefore: string;
      nonce: string;
    };
    signature: string;
  };
}

export type X402SignedPayload = X402SignedAuthorization | X402SignedAuthorizationV2;

export interface X402SignedAuthorization {
  x402Version: number;
  scheme: string;
  network: string;
  token: string;
  assetContract: string;
  authorization: {
    from: string;
    to: string;
    value: string;
    validAfter: number;
    validBefore: number;
    nonce: string;
    v: number;
    r: string;
    s: string;
    signature: string;
  };
}

export function parsePriceToUnits(priceStr: string | number, decimals: number = 6): bigint {
  if (typeof priceStr === 'number') return BigInt(Math.round(priceStr * 10 ** decimals));
  const clean = String(priceStr).replace(/[^0-9.]/g, '');
  const [whole, fraction = ''] = clean.split('.');
  const paddedFraction = (fraction + '0'.repeat(decimals)).slice(0, decimals);
  return BigInt(whole || '0') * BigInt(10 ** decimals) + BigInt(paddedFraction);
}

export function parsePaymentHeader(value: string | null): any {
  if (!value) return null;
  try {
    return JSON.parse(value);
  } catch {
    try {
      const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
      const padded = normalized + '='.repeat((4 - normalized.length % 4) % 4);
      if (typeof atob === 'function') {
        return JSON.parse(atob(padded));
      }
      return JSON.parse(Buffer.from(padded, 'base64').toString('utf8'));
    } catch {
      return null;
    }
  }
}

export class X402SignerClient {
  public readonly wallet: any;
  public readonly baseUrl: string;
  public readonly timeoutMs: number;
  public readonly expectedRecipient: string;
  public readonly maxPriceUsd: number;
  public readonly maxAmountUnits: bigint;

  constructor(options: X402SignerClientOptions = {}) {
    this.wallet = options.wallet || options.walletSigner || null;
    this.baseUrl = (options.baseUrl || 'https://api.m2msentinel.com').replace(/\/+$/, '');
    this.timeoutMs = Number(options.timeoutMs || 30000);
    this.expectedRecipient = EXPECTED_PAYOUT_RECIPIENT;
    this.maxPriceUsd = options.maxPriceUsd !== undefined ? Number(options.maxPriceUsd) : DEFAULT_MAX_PRICE_USD;
    this.maxAmountUnits = parsePriceToUnits(this.maxPriceUsd, 6);
  }

  async signPaymentAuthorization(challenge: X402PaymentChallenge = {}): Promise<X402SignedPayload> {
    if (!this.wallet) {
      throw new Error('Signer wallet is required to sign x402 payment authorization');
    }

    // 1. IMMUTABLE LOCAL SECURITY CONSTANTS (Never challenge-controlled)
    const chainId = BASE_CHAIN_ID; // Base Mainnet (8453)
    const tokenContract = BASE_USDC_CONTRACT; // Base USDC (0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913)
    const payTo = EXPECTED_PAYOUT_RECIPIENT; // M2M Sentinel Payout (0x6d6c398390cfb88f1cd42715b84906a0bd6652aa)

    // x402 v2 nests the offer under `accepts` and is identified by its presence.
    // Reading only the flat shape made every guard below miss silently: they
    // compared undefined and passed, and the amount fell back to a hardcoded
    // default, signing for less than the server demanded.
    const isV2 = Array.isArray(challenge.accepts) && challenge.accepts.length > 0;
    const offer: any = isV2 ? challenge.accepts![0] : challenge;

    const offeredChainId = typeof offer.network === 'string' && offer.network.startsWith('eip155:')
      ? Number(offer.network.slice('eip155:'.length))
      : (offer.chainId || challenge.chainId);
    // v2 carries the address in `asset`; the flat shape carries a symbol there
    // and the address in `assetContract`. Compare addresses to addresses only.
    const offeredAsset = [offer.assetContract, challenge.assetContract, offer.asset, challenge.asset]
      .find((value: any) => typeof value === 'string' && /^0x[0-9a-fA-F]{40}$/.test(value)) || null;
    const offeredPayTo = offer.payTo || offer.recipient || challenge.payTo;
    const offeredTokenName = (offer.extra && offer.extra.name) || challenge.tokenName;
    const offeredTokenVersion = (offer.extra && offer.extra.version) || challenge.tokenVersion;

    // 2. STRICT CHALLENGE INTEGRITY CHECKS (Refuse if challenge alters network, asset, or recipient)
    if (offeredChainId && Number(offeredChainId) !== BASE_CHAIN_ID) {
      throw new Error(`[x402 Security Policy] Refusing to sign on unverified network chainId: ${offeredChainId}. Autonomous signer strictly requires Base Mainnet (8453).`);
    }
    if (offeredAsset && offeredAsset.toLowerCase() !== BASE_USDC_CONTRACT.toLowerCase()) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unapproved asset: ${offeredAsset}. Autonomous signer strictly requires Base USDC (${BASE_USDC_CONTRACT}).`);
    }
    if (offeredPayTo && offeredPayTo.toLowerCase() !== EXPECTED_PAYOUT_RECIPIENT.toLowerCase()) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected recipient: ${offeredPayTo}. Autonomous signer strictly requires ${EXPECTED_PAYOUT_RECIPIENT}.`);
    }
    if (offeredTokenName && offeredTokenName !== 'USD Coin') {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected tokenName: ${offeredTokenName}. Expected USD Coin.`);
    }
    if (offeredTokenVersion && offeredTokenVersion !== '2') {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected tokenVersion: ${offeredTokenVersion}. Expected 2.`);
    }

    // 3. STRICT LOCAL PRICE CEILING CHECK
    // Never invent a price. A guessed amount produces an authorization the
    // server refuses, which is indistinguishable from a rejected payment.
    const rawAmount = offer.amount ?? offer.maxAmountRequired ??
      challenge.maxAmountRequired ?? challenge.amountUnits ?? challenge.amount ?? challenge.price;
    let requestedAmountUnits: string | null = null;
    if (typeof rawAmount === 'number' && Number.isInteger(rawAmount) && rawAmount > 0) {
      requestedAmountUnits = String(rawAmount);
    } else if (typeof rawAmount === 'string' && /^[0-9]+$/.test(rawAmount.trim())) {
      requestedAmountUnits = rawAmount.trim();
    } else if (typeof rawAmount === 'string' && /[$.]/.test(rawAmount)) {
      requestedAmountUnits = parsePriceToUnits(rawAmount, 6).toString();
    }
    if (!requestedAmountUnits || !/^[0-9]+$/.test(requestedAmountUnits) || BigInt(requestedAmountUnits) <= BigInt(0)) {
      throw new Error('[x402] The challenge carried no readable amount. Refusing to guess a price.');
    }
    if (BigInt(requestedAmountUnits) > this.maxAmountUnits) {
      throw new Error(`[x402 Security Policy] Requested amount (${requestedAmountUnits} units) exceeds local client authorized price ceiling (${this.maxAmountUnits.toString()} units / $${this.maxPriceUsd}).`);
    }
    const amountUnits = requestedAmountUnits;

    const now = Math.floor(Date.now() / 1000);
    // v2 servers bound the window with `maxTimeoutSeconds`; a longer window
    // than advertised is not more permissive, just outside what settles.
    const validAfter: any = isV2 ? '0' : now - 60;
    const validBefore: any = isV2 ? String(now + (Number(offer.maxTimeoutSeconds) || 120)) : now + 3600;
    
    // Generate 32-byte hex nonce
    let nonce = '0x';
    if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
      const bytes = new Uint8Array(32);
      crypto.getRandomValues(bytes);
      nonce += Array.from(bytes, b => b.toString(16).padStart(2, '0')).join('');
    } else {
      const cryptoNode = require('crypto');
      nonce += cryptoNode.randomBytes(32).toString('hex');
    }

    const domain = {
      name: 'USD Coin',
      version: '2',
      chainId,
      verifyingContract: tokenContract
    };

    const types = {
      TransferWithAuthorization: [
        { name: 'from', type: 'address' },
        { name: 'to', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'validAfter', type: 'uint256' },
        { name: 'validBefore', type: 'uint256' },
        { name: 'nonce', type: 'bytes32' }
      ]
    };

    const fromAddress = typeof this.wallet.getAddress === 'function' 
      ? await this.wallet.getAddress() 
      : (this.wallet.address || this.wallet.account?.address);

    const message = {
      from: fromAddress,
      to: payTo,
      value: amountUnits.toString(),
      validAfter,
      validBefore,
      nonce
    };

    let signature: string | { r: string; s: string; v: number };
    if (typeof this.wallet.signTypedData === 'function') {
      signature = await this.wallet.signTypedData(domain, types, message);
    } else if (typeof this.wallet._signTypedData === 'function') {
      signature = await this.wallet._signTypedData(domain, types, message);
    } else if (typeof this.wallet.signTypedDataV4 === 'function') {
      signature = await this.wallet.signTypedDataV4({ domain, types, message, primaryType: 'TransferWithAuthorization' });
    } else {
      throw new Error('Wallet does not implement EIP-712 signTypedData');
    }

    let v: number, r: string, s: string;
    if (typeof signature === 'string') {
      const clean = signature.startsWith('0x') ? signature.slice(2) : signature;
      r = '0x' + clean.slice(0, 64);
      s = '0x' + clean.slice(64, 128);
      v = parseInt(clean.slice(128, 130), 16);
      if (v < 27) v += 27;
    } else {
      r = signature.r;
      s = signature.s;
      v = signature.v;
    }

    const flatSignature = typeof signature === 'string' 
      ? signature 
      : '0x' + r.slice(2) + s.slice(2) + v.toString(16).padStart(2, '0');

    if (isV2) {
      // The canonical v2 envelope. Omitting `accepted` does not read as "no
      // payment": it crashes the server's offer matcher, which then answers 402
      // with an opaque body. There is deliberately no top-level scheme/network
      // in v2 -- both live inside `accepted`.
      return {
        x402Version: 2,
        resource: challenge.resource,
        accepted: offer,
          // Echo the server's advertised extensions back. The canonical client
          // merges paymentRequired.extensions into the payload; omitting the
          // field still settles the payment, so this failed silently -- but the
          // CDP Bazaar indexes a resource from the bazaar extension carried by
          // the settlement, so a payment without it is invisible to discovery.
        extensions: challenge.extensions,
        payload: {
          authorization: {
            from: fromAddress,
            to: payTo,
            value: amountUnits.toString(),
            validAfter: String(validAfter),
            validBefore: String(validBefore),
            nonce
          },
          signature: flatSignature
        }
      };
    }

    // Legacy flat envelope, for servers that do not advertise `accepts`.
    return {
      x402Version: 2,
      scheme: 'eip3009',
      network: `eip155:${chainId}`,
      token: tokenContract,
      assetContract: tokenContract,
      authorization: {
        from: fromAddress,
        to: payTo,
        value: amountUnits.toString(),
        validAfter,
        validBefore,
        nonce,
        v,
        r,
        s,
        signature: flatSignature
      }
    };
  }

  async fetchWithAutoPayment(path: string, options: any = {}): Promise<any> {
    const url = path.startsWith('http') ? path : `${this.baseUrl}/${path.replace(/^\/+/, '')}`;
    const headers = { 'Accept': 'application/json', 'User-Agent': '@m2msentinel/sdk-ts/1.2.3', ...(options.headers || {}) };

    let res = await fetch(url, { method: options.method || 'GET', headers, body: options.body });
    if (res.status !== 402) {
      return res;
    }

    const challengeHeader = res.headers.get('PAYMENT-REQUIRED') || res.headers.get('x402-payment-required');
    let challenge: any = parsePaymentHeader(challengeHeader);

    if (!challenge) {
      try {
        const bodyJson = await res.clone().json();
        // Pass the whole PaymentRequired, not just accepts[0]: the signer
        // needs `resource` and must echo the offer back as `accepted`.
        challenge = bodyJson.accepts ? bodyJson : bodyJson.paymentRequired;
      } catch {
        // Body was not JSON
      }
    }

    if (!challenge) {
      throw new Error('HTTP 402 received but no valid x402 payment challenge was found in headers or response body.');
    }

    const paymentPayload = await this.signPaymentAuthorization(challenge);
    const paymentB64 = typeof btoa === 'function'
      ? btoa(JSON.stringify(paymentPayload))
      : Buffer.from(JSON.stringify(paymentPayload)).toString('base64');

    const retryHeaders = {
      ...headers,
      'PAYMENT-SIGNATURE': paymentB64,
      'x402-payment-authorization': paymentB64
    };

    return fetch(url, { method: options.method || 'GET', headers: retryHeaders, body: options.body });
  }
}

export function x402SignerClient(options: X402SignerClientOptions = {}): X402SignerClient {
  return new X402SignerClient(options);
}

File: python/m2m_sentinel/x402.py
"""
M2M Sentinel — Autonomous Headless x402 Python Signer Client.

Handles HTTP 402 Payment Required challenges by signing EIP-712 / EIP-3009
transfer authorizations for Base USDC without browser wallet dependencies.
"""

import json
import base64
import time
import os
import re
import urllib.request
import urllib.error

BASE_USDC_CONTRACT = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
BASE_CHAIN_ID = 8453


EXPECTED_PAYOUT_RECIPIENT = "0x6d6c398390cfb88f1cd42715b84906a0bd6652aa"
DEFAULT_MAX_PRICE_USD = 0.05  # 5 cents maximum per autonomous request


class X402SignerClient:
    """Headless x402 payment client for autonomous Python agents on Base."""

    def __init__(self, private_key=None, base_url="https://api.m2msentinel.com", timeout=30, max_price_usd=DEFAULT_MAX_PRICE_USD):
        self.private_key = private_key or os.getenv("M2M_AGENT_WALLET_KEY")
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_price_usd = float(max_price_usd)
        self.max_amount_units = int(self.max_price_usd * 1_000_000)
        self.expected_recipient = EXPECTED_PAYOUT_RECIPIENT

    def _parse_challenge(self, header_value, body=None):
        if header_value:
            try:
                return json.loads(header_value)
            except Exception:
                try:
                    return json.loads(base64.b64decode(header_value).decode("utf-8"))
                except Exception:
                    pass
        if isinstance(body, dict):
            # Return the whole PaymentRequired, not accepts[0]: the signer needs
            # `resource` and must echo the chosen offer back as `accepted`.
            if "accepts" in body and len(body["accepts"]) > 0:
                return body
            if "paymentRequired" in body:
                return body["paymentRequired"]
        return None

    def sign_authorization(self, challenge=None):
        if not self.private_key:
            raise ValueError("Signer private key is required to sign x402 payment authorization")

        if challenge is None:
            challenge = {}

        # 1. IMMUTABLE LOCAL SECURITY CONSTANTS (Never challenge-controlled)
        chain_id = BASE_CHAIN_ID  # Base Mainnet (8453)
        token_contract = BASE_USDC_CONTRACT  # Base USDC (0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913)
        pay_to = EXPECTED_PAYOUT_RECIPIENT  # M2M Sentinel Payout (0x6d6c398390cfb88f1cd42715b84906a0bd6652aa)

        # x402 v2 nests the offer under `accepts` and is identified by its
        # presence. Reading only the flat shape made every guard below miss
        # silently -- they compared None and passed -- and the amount fell back
        # to a hardcoded default, signing for less than the server demanded.
        accepts = challenge.get("accepts")
        is_v2 = isinstance(accepts, list) and len(accepts) > 0
        offer = accepts[0] if is_v2 else challenge

        offered_network = offer.get("network") or challenge.get("network")
        if isinstance(offered_network, str) and offered_network.startswith("eip155:"):
            offered_chain_id = offered_network.split(":", 1)[1]
        else:
            offered_chain_id = offer.get("chainId") or challenge.get("chainId")

        # v2 carries the address in `asset`; the flat shape carries a symbol
        # there and the address in `assetContract`. Compare addresses only.
        offered_asset = None
        for candidate in (offer.get("assetContract"), challenge.get("assetContract"),
                          offer.get("asset"), challenge.get("asset")):
            if isinstance(candidate, str) and re.match(r"^0x[0-9a-fA-F]{40}$", candidate):
                offered_asset = candidate
                break

        offered_pay_to = offer.get("payTo") or offer.get("recipient") or challenge.get("payTo")
        extra = offer.get("extra") or {}
        offered_token_name = extra.get("name") or challenge.get("tokenName")
        offered_token_version = extra.get("version") or challenge.get("tokenVersion")

        # 2. STRICT CHALLENGE INTEGRITY CHECKS (Refuse if challenge alters network, asset, or recipient)
        if offered_chain_id and int(offered_chain_id) != BASE_CHAIN_ID:
            raise ValueError(f"[x402 Security Policy] Refusing to sign on unverified network chainId: {offered_chain_id}. Autonomous signer strictly requires Base Mainnet (8453).")
        if offered_asset and offered_asset.lower() != BASE_USDC_CONTRACT.lower():
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unapproved asset: {offered_asset}. Autonomous signer strictly requires Base USDC ({BASE_USDC_CONTRACT}).")
        if offered_pay_to and offered_pay_to.lower() != EXPECTED_PAYOUT_RECIPIENT.lower():
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unexpected recipient: {offered_pay_to}. Autonomous signer strictly requires {EXPECTED_PAYOUT_RECIPIENT}.")
        if offered_token_name and offered_token_name != "USD Coin":
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unexpected tokenName: {offered_token_name}. Expected USD Coin.")
        if offered_token_version and offered_token_version != "2":
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unexpected tokenVersion: {offered_token_version}. Expected 2.")

        # 3. STRICT LOCAL PRICE CEILING CHECK
        # Never invent a price. A guessed amount produces an authorization the
        # server refuses, which is indistinguishable from a rejected payment.
        raw_amount = (offer.get("amount") or offer.get("maxAmountRequired") or
                      challenge.get("maxAmountRequired") or challenge.get("amountUnits"))
        requested_amount_units = None
        if isinstance(raw_amount, int) and raw_amount > 0:
            requested_amount_units = str(raw_amount)
        elif isinstance(raw_amount, str) and raw_amount.strip().isdigit():
            requested_amount_units = raw_amount.strip()
        if not requested_amount_units or int(requested_amount_units) <= 0:
            raise ValueError("[x402] The challenge carried no readable amount. Refusing to guess a price.")
        if int(requested_amount_units) > self.max_amount_units:
            raise ValueError(f"[x402 Security Policy] Requested amount ({requested_amount_units} units) exceeds local client authorized price ceiling ({self.max_amount_units} units / ${self.max_price_usd}).")
        amount_units = requested_amount_units

        now = int(time.time())
        # v2 servers bound the window with `maxTimeoutSeconds`; a longer window
        # than advertised is not more permissive, just outside what settles.
        valid_after = 0 if is_v2 else now - 60
        valid_before = (now + int(offer.get("maxTimeoutSeconds") or 120)) if is_v2 else now + 3600
        nonce = "0x" + os.urandom(32).hex()

        try:
            from eth_account import Account
            from eth_account.messages import encode_typed_data

            account = Account.from_key(self.private_key)
            from_address = account.address

            typed_data = {
                "types": {
                    "EIP712Domain": [
                        {"name": "name", "type": "string"},
                        {"name": "version", "type": "string"},
                        {"name": "chainId", "type": "uint256"},
                        {"name": "verifyingContract", "type": "address"},
                    ],
                    "TransferWithAuthorization": [
                        {"name": "from", "type": "address"},
                        {"name": "to", "type": "address"},
                        {"name": "value", "type": "uint256"},
                        {"name": "validAfter", "type": "uint256"},
                        {"name": "validBefore", "type": "uint256"},
                        {"name": "nonce", "type": "bytes32"},
                    ],
                },
                "primaryType": "TransferWithAuthorization",
                "domain": {
                    "name": offered_token_name or "USD Coin",
                    "version": offered_token_version or "2",
                    "chainId": chain_id,
                    "verifyingContract": token_contract,
                },
                "message": {
                    "from": from_address,
                    "to": pay_to,
                    "value": int(amount_units),
                    "validAfter": valid_after,
                    "validBefore": valid_before,
                    "nonce": bytes.fromhex(nonce[2:]),
                },
            }

            signable = encode_typed_data(full_message=typed_data)
            signed = Account.sign_message(signable, private_key=self.private_key)

            signature_hex = signed.signature.hex()
            if not signature_hex.startswith("0x"):
                signature_hex = "0x" + signature_hex

            if is_v2:
                # The canonical v2 envelope. Omitting `accepted` does not read
                # as "no payment": it crashes the server's offer matcher, which
                # then answers 402 with an opaque body. There is deliberately no
                # top-level scheme/network in v2 -- both live inside `accepted`.
                return {
                    "x402Version": 2,
                    "resource": challenge.get("resource"),
                    "accepted": offer,
                    # Echo the server's advertised extensions back. The canonical
                    # client merges them into the payload; omitting the field
                    # still settles the payment, so this failed silently -- but
                    # the CDP Bazaar indexes a resource from the bazaar extension
                    # carried by the settlement, so a payment without it is
                    # invisible to discovery.
                    "extensions": challenge.get("extensions"),
                    "payload": {
                        "authorization": {
                            "from": from_address,
                            "to": pay_to,
                            "value": amount_units,
                            "validAfter": str(valid_after),
                            "validBefore": str(valid_before),
                            "nonce": nonce,
                        },
                        "signature": signature_hex,
                    },
                }

            # Legacy flat envelope, for servers that do not advertise `accepts`.
            return {
                "x402Version": 2,
                "scheme": "eip3009",
                "network": f"eip155:{chain_id}",
                "token": token_contract,
                "authorization": {
                    "from": from_address,
                    "to": pay_to,
                    "value": amount_units,
                    "validAfter": valid_after,
                    "validBefore": valid_before,
                    "nonce": nonce,
                    "v": signed.v,
                    "r": "0x" + hex(signed.r)[2:].zfill(64),
                    "s": "0x" + hex(signed.s)[2:].zfill(64),
                },
            }
        except ImportError:
            raise RuntimeError("eth-account is required for local EIP-712 signing: pip install eth-account")

    def fetch_with_auto_payment(self, path, method="GET", headers=None, body=None):
        url = path if path.startswith("http") else f"{self.base_url}/{path.lstrip('/')}"
        req_headers = {"Accept": "application/json", "User-Agent": "M2M-Sentinel-Python-Signer/1.2.3"}
        if headers:
            req_headers.update(headers)

        data = json.dumps(body).encode("utf-8") if body else None
        req = urllib.request.Request(url, data=data, headers=req_headers, method=method)

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read().decode("utf-8")
                try:
                    return {"status": resp.status, "data": json.loads(raw)}
                except Exception:
                    return {"status": resp.status, "text": raw}
        except urllib.error.HTTPError as err:
            if err.code != 402:
                raise

            challenge_header = err.headers.get("PAYMENT-REQUIRED") or err.headers.get("x402-payment-required")
            try:
                body_json = json.loads(err.read().decode("utf-8"))
            except Exception:
                body_json = None

            challenge = self._parse_challenge(challenge_header, body_json)
            if not challenge:
                raise ValueError("Could not extract x402 payment challenge from HTTP 402 response")

            payment_payload = self.sign_authorization(challenge)
            payment_b64 = base64.b64encode(json.dumps(payment_payload).encode("utf-8")).decode("utf-8")

            req_headers["PAYMENT-SIGNATURE"] = payment_b64
            retry_req = urllib.request.Request(url, data=data, headers=req_headers, method=method)

            with urllib.request.urlopen(retry_req, timeout=self.timeout) as resp:
                raw = resp.read().decode("utf-8")
                try:
                    return {"status": resp.status, "data": json.loads(raw)}
                except Exception:
                    return {"status": resp.status, "text": raw}
